rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,815 @@
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        feature_type: Feature type to be processed -- `sequence` or `sentence`.
+        feature_type_signature: A list of `FeatureSignature`s for the given attribute and
+            feature type.
+        config: A model config for correctly parametrising the layer.
+
+    Input shape:
+        List of N-D tensors, each with shape: `(batch_size, ..., input_dim)`.
+        All tensors must have the same shape, except the last dimension.
+        All sparse tensors must have the same shape including the last dimension.
+
+    Output shape:
+        N-D tensor with shape: `(batch_size, ..., units)` where `units` is the sum of
+        the last dimension sizes across all input sensors, with sparse tensors instead
+        contributing `config[DENSE_DIMENSION][attribute]` units each.
+
+    Raises:
+        ValueError if no feature signatures are provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        feature_type: Text,
+        feature_type_signature: List[FeatureSignature],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not feature_type_signature:
+            raise ValueError(
+                "The feature type signature must contain some feature signatures."
+            )
+
+        super().__init__(
+            name=f"concatenate_sparse_dense_features_{attribute}_{feature_type}"
+        )
+
+        self.output_units = self._calculate_output_units(
+            attribute, feature_type_signature, config
+        )
+
+        # Prepare dropout and sparse-to-dense layers if any sparse tensors are expected
+        self._tf_layers = {}
+        if any([signature.is_sparse for signature in feature_type_signature]):
+            self._prepare_layers_for_sparse_tensors(attribute, feature_type, config)
+
+    def _prepare_layers_for_sparse_tensors(
+        self, attribute: Text, feature_type: Text, config: Dict[Text, Any],
+    ) -> None:
+        """Set up pre-processing sparse tensors before combining them with dense ones."""
+        # For optionally applying dropout to sparse tensors
+        if config[SPARSE_INPUT_DROPOUT]:
+            self._tf_layers["sparse_dropout"] = layers.SparseDropout(
+                rate=config[DROP_RATE]
+            )
+
+        # For converting sparse tensors to dense
+        self._tf_layers["sparse_to_dense"] = layers.DenseForSparse(
+            name=f"sparse_to_dense.{attribute}_{feature_type}",
+            units=config[DENSE_DIMENSION][attribute],
+            reg_lambda=config[REGULARIZATION_CONSTANT],
+        )
+
+        # For optionally apply dropout to sparse tensors after they're converted to
+        # dense tensors.
+        if config[DENSE_INPUT_DROPOUT]:
+            self._tf_layers["dense_dropout"] = tf.keras.layers.Dropout(
+                rate=config[DROP_RATE]
+            )
+
+    def _calculate_output_units(
+        self,
```

Comment:
```suggestion
    @staticmethod
    def _calculate_output_units(
```

Reply:
Should all methods that don't use `self` be static?

Reply:
yes